### PR TITLE
spanconfigkvsubscriber,kvserver: fix KVSubscriber bug

### PR DESCRIPTION
**Pull request by arulajmani**
We had a bug in the KVSubscriber where we were invoking a copy of the
handler instead of the handler stored. This meant that we'd never treat
handlers as "initialized". As a result, we would always invoke them with
the everything span, and as a result, visit all replicas on the stores
in reaction to span config updates. See datadriven test diffs for an
illustration.

Fixing the above lead to unearthing an interesting bug in how we were
deciding to enqueue replicas in the split queue. Previously, if we
received  a span config update that implied a split and the update
corresponded to the right-hand side post split, we would skip enqueuing
the replica in the split queue. The assumption was that we'd get an
update corresponding to the LHS of the split for the same replica
and that update would enqueue the replica. This doesn't always hold true
though. For example, consider the case when a new table is created and
must be split from its (left) adjacent table's range. This only results
in a single update, corresponding to the new table's span, which is the
right-hand side post split.

This patch moves to nudging the split queue for all updates, not just
left-hand side updates, for the reason above.

Release note: None
Release justification: bug fixes in new functionality

### CODE REVIEW

**Review by nvb — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77064)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @irfansharif)

___
*[pkg/kv/kvserver/store.go, line 2432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77064#-Mx11F_81Fe378qxu3oj:-Mx11F_81Fe378qxu3ok:b-ckklhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/535a9603983f31484db89c5c2133479b0d99d9d4/pkg/kv/kvserver/store.go#L2432)):*
> ```Go
> 				}
> 				repl.SetSpanConfig(conf)
> 				return nil
> ```

Do we not want to enqueue the replica with the new span config in the two queues?
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go, line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77064#-Mx1-beZChyfweaDIrgk:-Mx1-beZChyfweaDIrgl:b-9rkdjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/535a9603983f31484db89c5c2133479b0d99d9d4/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L254)):*
> ```Go
> 		for _, ev := range events {
> 			target := ev.(*bufferEvent).Update.Target
> 			handlers[i].invoke(ctx, target.KeyspaceTargeted())
> ```

Nice catch! Do we have this issue in `handleCompleteUpdate` as well?

Also, consider rewriting this fix slightly to make it more explicit:
```go
for i := range handlers {
    handler := &handlers[i] // mutated by invoke
    for _, ev := range events {
        target := ev.(*bufferEvent).Update.Target
        handler.invoke(ctx, target.KeyspaceTargeted())
    }
}
```

**Review by arulajmani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77064)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/kv/kvserver/store.go, line 2432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77064#-Mx11F_81Fe378qxu3oj:-Mx17d837u6jN8IRodVC:bx5ffhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/535a9603983f31484db89c5c2133479b0d99d9d4/pkg/kv/kvserver/store.go#L2432)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we not want to enqueue the replica with the new span config in the two queues?
</blockquote></details>

We do, I'm not sure how I ended up pushing this `return nil` out -- I don't think it was intentional. 
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go, line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77064#-Mx1-beZChyfweaDIrgk:-Mx173z18MUm1RIW4H2z:bl0uy37) ([raw file](https://github.com/cockroachdb/cockroach/blob/535a9603983f31484db89c5c2133479b0d99d9d4/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L254)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Nice catch! Do we have this issue in `handleCompleteUpdate` as well?

Also, consider rewriting this fix slightly to make it more explicit:
```go
for i := range handlers {
    handler := &handlers[i] // mutated by invoke
    for _, ev := range events {
        target := ev.(*bufferEvent).Update.Target
        handler.invoke(ctx, target.KeyspaceTargeted())
    }
}
```
</blockquote></details>

Done. Also made the change to `handleCompleteUpdate` as well.

**Review by nvb — APPROVED** [review]
:lgtm: though @irfansharif should give this a look as well.

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77064#-:-Mx1BiPd0KHpRG_hK01u:b-42sr0v)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif)

File: pkg/kv/kvserver/store.go [hunk 1]
```diff
@@ -2395,45 +2395,43 @@ func (s *Store) onSpanConfigUpdate(ctx context.Context, updated roachpb.Span) {
 				return nil // placeholder; ignore
 			}
 
+			replCtx := repl.AnnotateCtx(ctx)
```
Comment (irfansharif) [inline pkg/kv/kvserver/store.go:2398]: [nit] Move this inside the conditional?

**Review by irfansharif — APPROVED** [review]

**Comment by arulajmani** [comment]
TFTRs! The CI failures here are flakes, gonna merge. 

bors r+

**Comment by arulajmani** [comment]
Flakes, plus didn't realize we need release justification in the PR description as well. 

bors r+

**Pull request merged**